src/foam/nanos/auth/ServiceProviderAwareDAO.js

```diff
@@ -0,0 +1,67 @@
+/**
+ * @license
+ * Copyright 2019 The FOAM Authors. All Rights Reserved.
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+foam.CLASS({
+  package: 'foam.nanos.auth',
+  name: 'ServiceProviderAwareDAO',
+  extends: 'foam.nanos.auth.AuthorizationDAO',
+
+  documentation: 'Enforce Spid to that of context or system user',
+
+  javaImports: [
+    'foam.core.FObject',
+    'foam.nanos.auth.User',
+    'foam.nanos.logger.Logger',
+    'foam.util.SafetyUtil'
+  ],
+
+  properties: [
+    {
+      name: 'authorizer',
+      class: 'FObjectProperty',
+      of: 'foam.nanos.auth.Authorizer',
+      javaFactory: `
+    return new AuthorizableAuthorizer("spid.update.*");
+`
+    },
+    {
+      name: 'defaultSpid',
+      class: 'String',
+      value: 'nanopay'
+    }
+  ],
+
+  methods: [
+    {
+      name: 'put_',
+      javaCode: `
+    if ( ! ( obj instanceof ServiceProviderAware ) ) {
+      return super.put_(x, obj);
+    }
+
+    ServiceProviderAware sp = (ServiceProviderAware) obj;
+    ServiceProviderAware oldSp = (ServiceProviderAware) getDelegate().inX(x).find(sp.getSpid());
+
+    boolean isCreate = oldSp == null || SafetyUtil.isEmpty(sp.getSpid()) || SafetyUtil.isEmpty(oldSp.getSpid());
+
+    if ( isCreate ) {
+      User user = (User) x.get("user");
+      if ( user != null &&
+           ! SafetyUtil.isEmpty(user.getSpid()) ) {
+        sp.setSpid(user.getSpid());
+      } else {
+        sp.setSpid(getDefaultSpid());
+      }
+    } else if ( ! oldSp.getSpid().equals(sp.getSpid()) ) {
+      authorizer_.authorizeOnUpdate(x, (FObject) oldSp, obj);
```

Comment:
I think you could just replace this line with

```suggestion
      if ( ! ((AuthService) getAuth()).check(x, "spid.update.*") ) {
        throw new AuthorizationException("You do not have permission to update SPIDs.");
      }
```

and get rid of the `extends: AuthorizationDAO`.

Reply:
Yes, I could but isn't that the point of AuthorizationDAO?

Reply:
No, AuthorizationDAO does much more. It's given an Authorizer and applies it to objects in the DAO during CRUD operations. The methods on `Authorizer` are `authorizeOnCreate(X, FObject)`, `authorizeOnRead(X, FObject)`, `authorizeOnUpdate(X, FObject)`, and `authorizeOnDelete(X, FObject)`. If the Authorizer is a `StandardAuthorizer`, it'll check permissions. If it's an `AuthorizableAuthorizer`, it'll call the `Authorizable` methods on the objects, which may or may not do permission checks. It depends on how the object chose to implement the `Authorizable` interface.

Here you're using an `AuthorizableAuthorizer`. I think you might have had `StandardAuthorizer` in mind. Using a `StandardAuthorizer` with "spid" as its permission prefix would check the "spid.update.<spid>" permission on update, "spid.create.<spid>" permission on create, etc. It would apply it for all CRUD operations on the DAO. However, that's not what we want here. We only care about protecting the SPID property from unauthorized changes, not stopping users from making any updates/creates/reads/deletes at all if they lack the spid permission.

We really only care about 1 specific case here: updating a SPID on a SpidAware FObject. Therefore it makes more sense just to use 1 specific permission check.